### PR TITLE
Support equal NaNs in NullEquals binary operation

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Test change.
Reference #12266 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/binaryop/compiled/operation.cuh [hunk 1]
```diff
@@ -390,9 +390,11 @@ struct NullEquals {
...
   {
     output_valid = true;
-    if (!lhs_valid && !rhs_valid) return true;
-    if (lhs_valid && rhs_valid) return x == y;
-    return false;
+    if (!lhs_valid || !rhs_valid) return rhs_valid == lhs_valid;
+    if constexpr (cudf::is_floating_point<TypeLhs>() && cudf::is_floating_point<TypeRhs>()) {
+      if (isnan(x) && isnan(y)) return true;
+    }
+    return x == y;
```
Comment (shwina) [inline cpp/src/binaryop/compiled/operation.cuh:397]: I think doing this unconditionally will break us. We usually want `NaN`s to compare not equal:

```python
In [11]: s = cudf.Series([1.0, 2.0, np.nan], nan_as_null=False)

In [12]: s == s
Out[12]: 
0     True
1     True
2    False
dtype: bool
```

Except when using `.equals()`, when we want NaNs to compare equal:

```
# s.equals(s.copy()) should return [True, True, True]
```

Comment (davidwendt) [reply cpp/src/binaryop/compiled/operation.cuh:397]: Ok. I wasn't sure if coding this only in `NullsEquals` would you help here.

**Comment by revans2** [comment]
I think this is fine from the spark side. Everywhere I looked that we use NULL_EQUALS it is followed by special case handling to make NaNs equal. 

I would add that we have to do some special case processing for NaNs on all of the comparison operators, not just NULL_EQUALS. But I think those are more likely to cause issues if we don't do this carefully.

**Comment by davidwendt** [comment]
Closing this based on comment from Ashwin that this would break cudf.

**Pull request closed without merging**